pyro/infer/trace_crps.py

```diff
@@ -0,0 +1,176 @@
+import operator
+from collections import OrderedDict
+from functools import reduce
+
+import torch
+
+import pyro
+import pyro.poutine as poutine
+from pyro.distributions.util import scale_and_mask
+from pyro.infer.elbo import ELBO
+from pyro.infer.util import is_validation_enabled
+from pyro.poutine.util import prune_subsample_sites
+from pyro.util import check_model_guide_match, check_site_shape, warn_if_nan
+
+
+def _squared_error(x, y, scale, mask):
+    diff = x - y
+    error = torch.einsum("np,np->n", diff, diff)
+    return scale_and_mask(error, scale, mask)
+
+
+class Trace_CRPS:
+    """
+    Posterior predictive CRPS loss with optional ``KL(q,p)`` regularization.
+
+    This is a likelihood-free method, and can be used for likelihoods without
+    tractible density functions. CRPS is a robust loss function, and is well
+    defined for any distribution with finite absolute moment ``E[|data|]``.
```

Comment:
data or latent?

Reply:
data; that is the space where we compute CRPS.